src/bootstrap/dist.rs

```diff
@@ -650,10 +655,14 @@ fn skip_host_target_lib(builder: &Builder<'_>, compiler: Compiler) -> bool {
 /// Copy stamped files into an image's `target/lib` directory.
 fn copy_target_libs(builder: &Builder<'_>, target: &str, image: &Path, stamp: &Path) {
     let dst = image.join("lib/rustlib").join(target).join("lib");
+    let self_contained_dst = dst.join("self_contained");
     t!(fs::create_dir_all(&dst));
+    t!(fs::create_dir_all(&self_contained_dst));
     for (path, dependency_type) in builder.read_stamp_file(stamp) {
-        if dependency_type != DependencyType::Host || builder.config.build == target {
+        if dependency_type == DependencyType::Target || builder.config.build == target {
             builder.copy(&path, &dst.join(path.file_name().unwrap()));
+        } else if dependency_type == DependencyType::SelfContained {
+            builder.copy(&path, &self_contained_dst.join(path.file_name().unwrap()));
```

Comment:
It seems like we would probably not want to special case the lack of cross-compilation and instead always place self-contained artifacts in the separate directory?

Reply:
Wow, totally missed that!
This should be either:
```rust
if dependency_type == DependencyType::SelfContained {
    builder.copy(&path, &self_contained_dst.join(path.file_name().unwrap()));
} else if dependency_type == DependencyType::Target || builder.config.build == target {
    builder.copy(&path, &dst.join(path.file_name().unwrap()));
}
```
or more explicit like this:
```rust
if dependency_type == DependencyType::Target
    || dependency_type == DependencyType::Host && builder.config.build == target
{
    builder.copy(&path, &dst.join(path.file_name().unwrap()));
} else if dependency_type == DependencyType::SelfContained {
    builder.copy(&path, &self_contained_dst.join(path.file_name().unwrap()));
}
```
What do you tink?

Reply:
I'd personally prefer the first one I think without reading too much into it, but I'm not sure why we're special casing the non-cross-compilation case. We do that quite a bit in bootstrap though unfortunately.

Reply:
> I'm not sure why we're special casing the non-cross-compilation case. We do that quite a bit in bootstrap though unfortunately.

I didn't want to break it so I just didn't touch that 😄 
Upon quick look at it only proc-macro dynamic libraries and their import libraries are marked as `Host` (maybe I should change the enum kind from `Host` to `ProcMacro`?).
Proc-macros always run on the host so if I'm not mistaken when cross compiling we would actually copy library built for our host instead of the target. So in the end we copy it only when they match the target (host == target) and skip them otherwise.

That's how I see it.